src/stacked_borrows/diagnostics.rs

```diff
@@ -0,0 +1,338 @@
+use rustc_middle::mir::interpret::{AllocId, AllocRange};
+use rustc_span::{Span, SpanData};
+use rustc_target::abi::Size;
+
+use crate::helpers::HexRange;
+use crate::stacked_borrows::{err_sb_ub, AccessKind, GlobalStateInner, Permission};
+use crate::Item;
+use crate::SbTag;
+use crate::Stack;
+use crate::ThreadManager;
+
+use rustc_middle::mir::interpret::InterpError;
+
+#[derive(Debug, Default)]
+pub struct AllocHistory {
+    // The time tags can be compressed down to one bit per event, by just storing a Vec<u8>
+    // where each bit is set to indicate if the event was a creation or a retag
+    current_time: usize,
+    creations: smallvec::SmallVec<[Event; 2]>,
+    invalidations: smallvec::SmallVec<[Event; 1]>,
+    protectors: smallvec::SmallVec<[Protection; 1]>,
+}
+
+#[derive(Debug)]
+struct Protection {
+    orig_tag: SbTag,
+    tag: SbTag,
+    span: Span,
+}
+
+#[derive(Debug)]
+struct Event {
+    time: usize,
+    parent: Option<SbTag>,
+    tag: SbTag,
+    range: AllocRange,
+    span: Span,
+}
+
+pub enum TagHistory {
+    Tagged {
+        tag: SbTag,
+        created: (AllocRange, SpanData),
+        invalidated: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+    Untagged {
+        recently_created: Option<(AllocRange, SpanData)>,
+        recently_invalidated: Option<(AllocRange, SpanData)>,
+        matching_created: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+}
+
+pub trait GlobalStateExt {
+    fn current_span(&self, threads: &ThreadManager<'_, '_>) -> Span;
+
+    fn log_creation(
+        &mut self,
+        parent: Option<SbTag>,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn log_invalidation(
+        &mut self,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn log_protector(
+        &mut self,
+        orig_tag: SbTag,
+        tag: SbTag,
+        alloc: AllocId,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn get_stack_history(
+        &self,
+        tag: SbTag,
+        alloc: AllocId,
+        alloc_range: AllocRange,
+        offset: Size,
+        protector_tag: Option<SbTag>,
+    ) -> Option<TagHistory>;
+}
+
+impl GlobalStateExt for GlobalStateInner {
+    fn current_span(&self, threads: &ThreadManager<'_, '_>) -> Span {
+        threads
```

Comment:
I think this should be a method on the ThreadManager. Maybe `local_crates` could even live in the `ThreadManager`.

Reply:
> Maybe local_crates could even live in the ThreadManager.

Hm, that seems like a bit of a hack but it would be rather convenient here so :shrug: